File: evaluator/src/main/java/reco/evaluator/evaluation/metrics/MRR.java
package reco.evaluator.evaluation.metrics;

import reco.evaluator.models.Recommendation;

import java.util.List;
import java.util.stream.Collectors;

public class MRR implements SessionMetric {

    private double sumOfScores = 0;
    private double qty = 0;

    private final int length;

    public MRR() {
        this(20);
    }

    public MRR(int length) {
        this.length = length;
    }


    @Override
    public void add(List<Recommendation> recommendations, List<Long> nextItems) {
        this.qty += 1;
        List<Long> recommendationsAsList = recommendations.subList(0, Math.min(this.length, recommendations.size())).
                stream().map(r -> r.getItemId()).collect(Collectors.toList());
        long nextItem = nextItems.get(0);

        int rank = recommendationsAsList.indexOf(nextItem);
        if (rank >= 0) {
            this.sumOfScores += 1 / (rank + 1.0);
        }
    }

    @Override
    public double result() {
        return this.sumOfScores / this.qty;
    }

    @Override
    public String getName() {
        return this.getClass().getSimpleName() + "@" + this.length;
    }
}

File: rust_optimized_impl/src/metrics/mrr.rs
use std::cmp;
use itertools::Itertools;
use crate::metrics::SessionMetric;


pub struct Mrr {
    sum_of_scores: f64,
    qty: usize,
    length: usize,
}

impl Mrr {
}

impl Mrr {
    pub fn new(length: usize) -> Mrr {
        Mrr {
            sum_of_scores: 0_f64,
            qty: 0,
            length: length,
        }
    }
}

impl SessionMetric for Mrr {
    fn add(&mut self, recommendations: &Vec<u64>, next_items: &Vec<u64>) {
        self.qty += 1;
        let top_recos = recommendations.iter().take(cmp::min(recommendations.len(), self.length)).collect_vec();
        let next_item = next_items[0];
        let index = top_recos.iter().position(|&&item_id| item_id == next_item);
        match index {
            Some(rank) => {
                self.sum_of_scores += 1_f64 / (rank as f64 + 1_f64)
            },
            None => (),
        }
    }

    fn result(&self) -> f64 {
        self.sum_of_scores / self.qty as f64
    }

    fn get_name(&self) -> String {
        format!("Mrr@{}", self.length)
    }

}